regression-tests/tests/adt-recursive.ssl

```diff
@@ -0,0 +1,49 @@
+type Tree
+  TwoChildren Tree Tree Int
+  OneChild Int Tree
+  Leaf
+ 
+//         A                         A ~ 65
+//       /   \                       B ~ 66
+//      B     C                      C ~ 67
+//     / \     \                     D ~ 68
+//    D   E     *                    E ~ 69
+//   /     \
+//  *       *      
+
+// 1) print out the left-most descendant
+// 2) print out the right-most descendant
+// expected output: DC
+
+main ( cout : & Int ) -> () =
+  let d = OneChild 68 Leaf
+  let e = OneChild 69 Leaf
+  let b = TwoChildren d e 66
+  let c = OneChild 67 Leaf
+  let root = TwoChildren b c 65
+  let lmd = match root
+                OneChild v kid = 88
+                TwoChildren l r v = match l 
+                                       OneChild _ _ = 89
+                                       Leaf = 87
+                                       TwoChildren k1 k2 val = match k1
+                                                                  OneChild v kid = match kid
+                                                                                        Leaf = v 
+                                                                                        _ = 88
+                                                                  _ = 88
+                                       _ = 88
+                _ = 88
```

Comment:
Note about writing pattern-matches: you should be able to save on some horizontal space by writing:

```
let x =
  match e
    ...
```

instead of

```
let x = match e
          ...
```

Reply:
I am actually unable to get a match like that to compile:
```
type Tree
  TwoChildren Tree Tree  
  Leaf                   
  OneChild Int Tree

//main ( cout : & Int ) -> () =
//  let d = Leaf 
//  let zzz = match d 
//                  Leaf = 65
//                  _ = 66

main ( cout : & Int ) -> () =
  let d = Leaf 
  let zzz = 
    match d 
          Leaf = 65
          _ = 66

// correct about should be 65 'A'
  after 10 , (cout : & Int) <- zzz
  wait (cout : & Int)
```
If I don't put the match on the same line as the assignment, I get the following error:
```
###### Testing match-indentation
stack exec sslc -- tests/match-indentation.ssl > out/match-indentation.c
ParseError (ErrorMsg "Cannot start block at lower indentation than before")
###### FAILED
```
I will make a bug issue about this.